src/compiler/checker.ts

```diff
@@ -1192,6 +1195,40 @@ namespace ts {
                     return false;
                 });
             }
+
+            function isPropertyImmediatelyReferencedWithinDeclaration(declaration: PropertyDeclaration, usage: Node) {
+                // always legal if usage is after declaration
+                if (usage.end > declaration.end) {
+                    return true;
+                }
+
+                // still might be legal if usage is deferred (e.g. x: any = () => this.x)
+                // otherwise illegal if immediately referenced within the declaration (e.g. x: any = this.x)
+                const ancestorChangingReferenceScope = findAncestor(usage, (node: Node) => {
+                    if (node === declaration) {
+                        return "quit";
+                    }
+
+                    switch (node.kind) {

```

Comment:
This is suspiciously similar to the change in #29374. That one searches downward, instead of upward, but has almost the same cases. It'd be nice if an improvement _(already-existing function ? re-using existing control flow analysis? that same function?)_ there could be used here.

Reply:
similar code is also used in 'used before declaration', 'used before being assigned' and 'parameter cannot reference identifier declared after it'.
all of them do it slightly different